tests/src/utils.rs

```diff
@@ -0,0 +1,236 @@
+use std::{collections::HashMap, env, panic, path::Path, process, process::Stdio};
+
+use k8s_openapi::{
+    api::{
+        apps::v1::Deployment,
+        core::v1::{Namespace, Pod, Service},
+    },
+    apimachinery::pkg::apis::meta::v1::ObjectMeta,
+};
+use kube::{
+    api::{DeleteParams, ListParams, PostParams},
+    Api, Client, Config,
+};
+use reqwest::{Method, StatusCode};
+use serde_json::json;
+use tokio::{
+    io::{AsyncBufReadExt, BufReader},
+    process::{Child, ChildStdout, Command},
+    time::{sleep, Duration},
+};
+
+static TEXT: &str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
+
+// target/debug/mirrord exec --pod-name pod_name  -c binary command
+pub fn start_node_server(pod_name: &str, command: Vec<&str>, env: HashMap<&str, &str>) -> Child {
+    let path = env!("CARGO_BIN_FILE_MIRRORD");
+    let args: Vec<&str> = vec!["exec", "--pod-name", pod_name, "-c"]
+        .into_iter()
+        .chain(command.into_iter())
+        .collect();
+    let server = Command::new(path)
+        .args(args)
+        .envs(&env)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()
+        .unwrap();
+    server
+}
+
+pub async fn setup_kube_client() -> Client {
+    let mut config = Config::infer().await.unwrap();
+    config.accept_invalid_certs = true;
+    let client = Client::try_from(config).unwrap();
+    client
+}
+
+// minikube service nginx --url
+pub async fn get_service_url(client: &Client, namespace: &str) -> Option<String> {
+    let service_api: Api<Service> = Api::namespaced(client.clone(), namespace);
+    let pod_api: Api<Pod> = Api::namespaced(client.clone(), namespace);
+    for pod in pod_api.list(&ListParams::default()).await.unwrap() {
+        if pod.metadata.name.unwrap().contains("nginx") {
```

Comment:
Rather than querying all the pods and filtering them in memory, you can filter them in the api call and then just call .first(). See pod_api.rs in mirrord_layer for an example.

Reply:
This can wait for after the merge.